### PR TITLE
Fix masked tensor test_stack memory leak

### PR DESCRIPTION
**Pull request by huydhn**
This test is currently failing in trunk when memory leak check is enabled, for example https://github.com/pytorch/pytorch/actions/runs/11296206361/job/31422348823#step:22:1970.  When testing locally, calling `backward` on a masked tensor always causes memory leak until I clean up the data and the mask manually.  This is probably related to this warning from masked tensor `UserWarning: It is not recommended to create a MaskedTensor with a tensor that requires_grad. To avoid this, you can use data.clone().detach()`, but I don't know much about the internal details here to go further.  So, let's just fix the test first/

### Testing

```
PYTORCH_TEST_CUDA_MEM_LEAK_CHECK=1 python test/test_maskedtensor.py TestBasicsCUDA.test_stack_cuda
```

passes and doesn't warn about memory leak anymore.

The test itself came from https://github.com/pytorch/pytorch/pull/125262#issuecomment-2344068012

### CODE REVIEW

**Review by kit1980 — APPROVED** [review]

**Comment by kit1980** [comment]
The warning for the requires_grad maybe unrelated and can be easily fixed by removing `requires_grad=True`.
But `requires_grad=True` seems to have a purpose there, I don't understand what exactly.

**Comment by huydhn** [comment]
@cpuhrsch I see that you are the reviewer of https://github.com/pytorch/pytorch/pull/125262#issuecomment-2344068012 and would appreciate if you have any insights to share here

**Comment by cpuhrsch** [comment]
@huydhn - I'm not sure why this is happening. The original author of MaskedTensor left and this is mostly in community mode. If the fix in this PR resolves the leak in the test I suggest we land it.

**Comment by cpuhrsch** [comment]
@huydhn - Thanks for landing the hotfix. I'll dig into this more tomorrow just in case it's a symptom of something worse. Thank you.

**Comment by nowtryz** [comment]
Hi, Thank you for the fix!

If requires_grad caused the memory leak the in the test, I can look into it and manually require the grad on the masked tensor.

@cpuhrsch I did not consider it but I was using masked tensors just for parts of my code so gradients were passing through the masked part. Do you know where the leak may come from? Otherwise we could manualy delete the tensors in `MaskedTensor.__del__`?

**Comment by huydhn** [comment]
Yeah, https://github.com/pytorch/pytorch/pull/137890 has the proper fix
